File: Path Sum II/Path Sum II.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a list of lists of integers
    
    def pathSum(self, root, sum):
        if root is None:
            return []
        elif root.left is None and root.right is None:
            if sum == 0:
                return [[root]]
            else:
                return []
        elif root.left is None:
            res = self.pathSum(root.right, sum - root.val)
            return [[root] + item for item in res]
        elif root.right is None:
            res = self.pathSum(root.left, sum - root.val)
            return [[root] + item for item in res]
        else:
            res = self.pathSum(root.left, sum - root.val)
            res.extend(self.pathSum(root.right, sum - root.val))
            return [[root] + item for item in res]
        
        

File: Path Sum II/Path Sum II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    
    private void recur(TreeNode root, List<Integer> solution, List<List<Integer>> ret, int target, int sum){
        if(root == null){
            return;
        }
        
        sum += root.val;
        
        if((root.left == null) && (root.right == null)){
            if(target == sum){
                solution.add(root.val);
                ret.add(new ArrayList<Integer>(solution));
                solution.remove(solution.size() - 1);
            }
            return;
        }
        
        solution.add(root.val);
        recur(root.left, solution, ret, target, sum);
        recur(root.right, solution, ret, target, sum);
        solution.remove(solution.size() - 1);
        
        return;
    }
    
    
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> ret = new ArrayList<List<Integer>>();
        List<Integer> sol = new ArrayList<Integer>();
        
        if(root != null){
            recur(root, sol, ret, sum, 0);
        }
        
        return ret;
    }
}
